source/cppfront.cpp

```diff
@@ -4537,67 +4548,111 @@ class cppfront
                     }
 
-                    //  If this is an initialization of an 'out' parameter, stash it
-                    if (n.has_out_parameter_named(lhs)){
-                        out_inits.push_back( print_to_string(**statement, false) );
-                        (*statement)->emitted = true;
-                        ++statement;
-                        continue;
-                    }
-
-                    //  Now we're ready to check whether this is an assignment to *object
-
+                    //  Now we're ready to check whether this is a member assignment
                     if (!lhs.empty())
                     {
-                        //  First, see if it's an assignment 'name = something'
-                        found_explicit_init = object_name == lhs;
-
-                        //  Otherwise, see if it's 'this.name = something'
-                        if (!found_explicit_init)
+                        //  First, check if it's 'this = that'
+                        if (lhs == "this" && rhs == (emitting_move_that_function ? "std::move(that)" : "that"))
                         {
-                            //  If it's of the form 'this.name', check 'name'
-                            if (
-                                starts_with( lhs, "(*this).")
-                                && object_name == lhs.substr(8)
-                                )
+                            if (!emitting_that_function)
+                            {
+                                errors.emplace_back(
+                                    stmt_pos,
+                                    "'this = that' can only be used in functions where 'that' is a parameter"
+                                );
+                                return;
+                            }
+                            else if (object != objects.begin())
                             {
-                                found_explicit_init = true;
+                                errors.emplace_back(
+                                    stmt_pos,
+                                    "'this = that' must appear before all other member initializers"
+                                );
+                                return;
+                            }
+                            else if (found_that_init)
+                            {
+                                errors.emplace_back(
+                                    stmt_pos,
+                                    "'this = that' may only appear once"
+                                );
+                                return;
+                            }
+                            else
+                            {
+                                // All the statements before 'this = that' need to be inserted
+                                // before the first member initializer whether or not it's explicit
+                                pending_statements = deferred_statements;
```

Comment:
This is still a good optimization to prevent allocation:

    pending_statements = std::move(deferred_statements);

You can also do it like this:

    std::swap(pending_statements, deferred_statements);
    deferred_statements.clear();